chrysom/client.go

```diff
@@ -129,140 +162,229 @@ func validateConfig(config *ClientConfig) error {
 	}
 	return nil
 }
+func shouldUseJWTAcquirer(options acquire.RemoteBearerTokenAcquirerOptions) bool {
+	return len(options.AuthURL) > 0 && options.Buffer != 0 && options.Timeout != 0
+}
 
-func determineTokenAcquirer(config ClientConfig) (acquire.Acquirer, error) {
-	defaultAcquirer := &acquire.DefaultAcquirer{}
-	if config.Auth.JWT.AuthURL != "" && config.Auth.JWT.Buffer != 0 && config.Auth.JWT.Timeout != 0 {
-		return acquire.NewRemoteBearerTokenAcquirer(config.Auth.JWT)
+func buildTokenAcquirer(auth *Auth) (acquire.Acquirer, error) {
+	if shouldUseJWTAcquirer(auth.JWT) {
+		return acquire.NewRemoteBearerTokenAcquirer(auth.JWT)
+	} else if len(auth.Basic) > 0 {
+		return acquire.NewFixedAuthAcquirer(auth.Basic)
 	}
+	return &acquire.DefaultAcquirer{}, nil
+}
 
-	if config.Auth.Basic != "" {
-		return acquire.NewFixedAuthAcquirer(config.Auth.Basic)
+func validateGetItemsInput(input *GetItemsInput) error {
+	if input == nil {
+		return ErrUndefinedInput
 	}
 
-	return defaultAcquirer, nil
+	if len(input.Bucket) < 1 {
+		return ErrBucketEmpty
+	}
+	return nil
 }
 
-func (c *Client) GetItems(owner string) ([]model.Item, error) {
-	request, err := http.NewRequest("GET", fmt.Sprintf("%s/api/v1/store/%s", c.remoteStoreAddress, c.bucketName), nil)
+func (c *Client) makeRequest(owner, method, URL string, body io.Reader) (*http.Request, error) {
+	r, err := http.NewRequest(method, URL, body)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrNewRequestFailure, err.Error())
 	}
-	err = acquire.AddAuth(request, c.auth)
+	err = acquire.AddAuth(r, c.auth)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("%w: %s", ErrAuthAcquirerFailure, err.Error())
+	}
+
+	if len(owner) > 0 {
+		r.Header.Set(store.ItemOwnerHeaderKey, owner)
+	}
+
+	return r, nil
+}
+
+func (c *Client) do(r *http.Request) (*doResponse, error) {
+	resp, err := c.client.Do(r)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrDoRequestFailure, err.Error())
 	}
+	defer resp.Body.Close()
 
-	request.Header.Set(store.ItemOwnerHeaderKey, owner)
+	var dResp = doResponse{
+		Code: resp.StatusCode,
+	}
+	bodyBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return &dResp, fmt.Errorf("%w: %s", ErrReadingBodyFailure, err.Error())
+	}
+	dResp.Body = bodyBytes
+	return &dResp, nil
+}
 
-	response, err := c.client.Do(request)
+// GetItems fetches all items in a bucket that belong to a given owner.
+func (c *Client) GetItems(input *GetItemsInput) (*GetItemsOutput, error) {
```

Comment:
I pretty much have the same issue with other `Input` structs.

Reply:
I followed a similar pattern to what folks at AWS use for their dynamodb go sdk https://docs.aws.amazon.com/sdk-for-go/api/service/dynamodb/#DynamoDB.PutItem

One big benefit is that it facilitates updating the client without changing the signature of the methods. We can deprecate new fields and add new ones in a backwards compatible way.

Even when our structs are not super large, `PushItem(*PushItemInput)PushItemOutput` already ran around 4000ns faster / op than `PushItemOutput(PushItemInput)PushItemOutput`

```
goos: darwin
goarch: amd64
pkg: github.com/xmidt-org/argus/chrysom
BenchmarkPushItemByValue-12    	   13467	     80400 ns/op	    7841 B/op	     100 allocs/op
PASS
ok  	github.com/xmidt-org/argus/chrysom	2.140s



goos: darwin
goarch: amd64
pkg: github.com/xmidt-org/argus/chrysom
BenchmarkPushItem-12    	   15087	     77748 ns/op	    7865 B/op	     101 allocs/op
PASS
ok  	github.com/xmidt-org/argus/chrysom	2.248s

```

Reply:
Everything I research about passing a pointer vs value says that value is faster:
https://medium.com/a-journey-with-go/go-should-i-use-a-pointer-instead-of-a-copy-of-my-struct-44b43b104963

I often find that pointers are used by SDKs that are written in multiple languages...it doesn't feel very Go-like to me.  Writing functions in this way seems very different from how we currently write code, and if we want to do this I think it should be a team decision.  Personally, I don't think the benefits of potential future backwards compatibility makes up for the potential for bugs, additional code maintenance, and larger mental model needed to understand the code (now I cannot simply look at the function signature and understand what's going on - I have to go look up what these structs are).  But if this is the direction the team wants to move in, then we can.

Reply:
> `*GetItemsOutput` is a pointer to a struct that only has a slice, which is a pointer. Why can't we just return the slice?

The decision here is more about how the API will evolve.  It feels to me like this method won't ever need to return anything but a slice, so have a full-on `struct` as a result type seems unnecessary.  I'm not as close to the design, though.

Don't forget that you can take the middle ground and return a custom slice type:

```
type Items []Item
func GetItemsOutput(...) Items
```

That has a few advantages:

(1) You can now introduce methods on your return type `Items`.  That helps the API evolve over time, as you can introduce additional behavior without having to change the `GetItemsOutput` function.

(2) You can document the `Items` type with `godoc`

(3) You can switch to a `type Items struct {...}` in the future without too much of a breaking change.  Since clients won't typically instantiate the `Items` type themselves, it won't break very much code.